components/fxa-client/src/internal/state_manager.rs

```diff
@@ -186,6 +207,31 @@ impl StateManager {
         self.persisted_state.refresh_token = Some(refresh_token);
         self.persisted_state.access_token_cache.clear();
         self.persisted_state.device_capabilities.clear();
+        self.save_state();
+    }
+
+    /// Save the state via our [StorageHandler], if one is registered
+    ///
+    /// This should be called after every &mut function that changes the persisted state
+    fn save_state(&self) {
+        if let Some(storage_handler) = &self.storage_handler {
+            match state_to_json(&self.persisted_state) {
+                Ok(json) => {
+                    if let Err(e) = storage_handler.save_state(json) {
+                        error_support::report_error!(
+                            "fxaclient-persist-state",
+                            "Error saving state: {e}"
+                        );
+                    }
```

Comment:
> How does this change error handling in persisting the state? Do we know how android and iOS react nowadays when the state persistence fails?

From what I can tell [Swift catches all errors and converts them to warnings](https://github.com/mozilla/application-services/blob/main/components/fxa-client/ios/FxAClient/PersistedFirefoxAccount.swift#L246-L258).  [Kotlin does that for JSON conversion errors, but lets other errors bubble up](https://github.com/mozilla/application-services/blob/main/components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/PersistedFirefoxAccount.kt#L77-L88)

> Another side effect here we should double check is that previously, the state was only ever updated once the high-level operation was fully completed, where now it can be partially updated (possibly multiple times) before the operation is complete.

That's a very good point.  Maybe we only emit one save state call per high-level function.  I think that would mean that saving a flag if the persistent state changed that we would then check at the top-level.  I think we could probably have a standard wrapper function for all top-level methods that did that.